lib/events/emitter.go

```diff
@@ -31,6 +32,108 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+// AsyncEmitterConfig provides parameters for emitter
+type AsyncEmitterConfig struct {
+	// Inner emits events to the underlying store
+	Inner Emitter
+	// BufferSize is a default buffer size for emitter
+	BufferSize int
+}
+
+// CheckAndSetDefaults checks and sets default values
+func (c *AsyncEmitterConfig) CheckAndSetDefaults() error {
+	if c.Inner == nil {
+		return trace.BadParameter("missing parameter Inner")
+	}
+	if c.BufferSize == 0 {
+		c.BufferSize = defaults.AsyncBufferSize
+	}
+	return nil
+}
+
+// NewAsyncEmitter returns emitter that submits events
+// without blocking the caller. It will start loosing events
+// on buffer overflow.
+func NewAsyncEmitter(cfg AsyncEmitterConfig) (*AsyncEmitter, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	completeCtx, completeSignal := context.WithCancel(context.Background())
+	a := &AsyncEmitter{
+		cancel:         cancel,
+		ctx:            ctx,
+		completeSignal: completeSignal,
+		completeCtx:    completeCtx,
+		eventsCh:       make(chan AuditEvent),
+		cfg:            cfg,
+	}
+	go a.forward()
+	return a, nil
+}
+
+// AsyncEmitter accepts events to a buffered channel and emits
+// events in a separate goroutine without blocking the caller.
+type AsyncEmitter struct {
+	cfg      AsyncEmitterConfig
+	eventsCh chan AuditEvent
+	cancel   context.CancelFunc
+	ctx      context.Context
+
+	completeSignal context.CancelFunc
+	completeCtx    context.Context
+}
+
+// Close closes emitter and cancels all in flight events.
+// It attempts to exit gracefully without loosing in-flight events,
+// but will return if the context has timed out
+func (a *AsyncEmitter) Close(ctx context.Context) error {
+	a.cancel()
+	select {
+	// wait for all in-flight emit calls to complete
+	case <-a.completeCtx.Done():
+		return nil
+	case <-ctx.Done():
+		return trace.ConnectionProblem(
+			ctx.Err(), "context was canceled or timed out before clean close could finish")
+	}
+	return nil
+}
+
+func (a *AsyncEmitter) forward() {
+	// signal that this goroutine has been completed
+	defer a.completeSignal()
+	for {
+		select {
+		case <-a.ctx.Done():
+			// in case if parent context closed,
+			// close the child context to avoid holding reference
+			// to the parent context
+			a.cancel()
```

Comment:
This should not be necessary - the only way `a.ctx` is closed is via `a.cancel`?